BUG: SciPy 1.7.x build failure on Cygwin

### Describe your issue.

SciPy 1.7.2 and 1.7.3 fail to build from source on Cygwin - https://trac.sagemath.org/ticket/33080

The failures come in through the vendored boost math library. as reported in https://github.com/boostorg/math/pull/727 and https://github.com/boostorg/math/pull/728


### Reproducing Code Example

```python
pip install scipy
```


### Error message

```shell
scipy/stats/_boost/src/beta_ufunc.cxx:3903:46:   required from here
  /opt/sage-4525b51c9c556e7d1bf7a4d90ee129c6b7f1d9a7/var/tmp/sage/build/scipy-1.7.3/src/scipy/_lib/boost/boost/math/tools/promotion.hpp:148:37: error: static assertion failed: Sorry, but this platform does not have sufficient long double support for the special functions to be reliably implemented.
    148 |          BOOST_STATIC_ASSERT_MSG((0 == ::boost::is_same<type, long double>::value), "Sorry, but this platform does not have sufficient long double support for the special functions to be reliably implemented.");
error: Command "g++ -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -ggdb -O2 -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector-strong --param=ssp-buffer-size=4 -DOPENSSL_NO_SSL3=1 -fdebug-prefix-map=/pub/devel/python/python38/python38-3.8.12-1.x86_64/build=/usr/src/debug/python38-3.8.12-1 -fdebug-prefix-map=/pub/devel/python/python38/python38-3.8.12-1.x86_64/src/Python-3.8.12=/usr/src/debug/python38-3.8.12-1 -ggdb -O2 -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector-strong --param=ssp-buffer-size=4 -DOPENSSL_NO_SSL3=1 -fdebug-prefix-map=/pub/devel/python/python38/python38-3.8.12-1.x86_64/build=/usr/src/debug/python38-3.8.12-1 -fdebug-prefix-map=/pub/devel/python/python38/python38-3.8.12-1.x86_64/src/Python-3.8.12=/usr/src/debug/python38-3.8.12-1 -D_GNU_SOURCE -DBOOST_MATH_DOMAIN_ERROR_POLICY=ignore_error -DBOOST_MATH_EVALUATION_ERROR_POLICY=user_error -DBOOST_MATH_PROMOTE_DOUBLE_POLICY=false -Iscipy/stats/_boost/include/ -Iscipy/stats/_boost/src/ -I/opt/sage-4525b51c9c556e7d1bf7a4d90ee129c6b7f1d9a7/lib/python3.8/site-packages/numpy/core/include -I/opt/sage-4525b51c9c556e7d1bf7a4d90ee129c6b7f1d9a7/var/tmp/sage/build/scipy-1.7.3/src/scipy/_lib/boost -I/opt/sage-4525b51c9c556e7d1bf7a4d90ee129c6b7f1d9a7/lib/python3.8/site-packages/numpy/core/include -Ibuild/src.cygwin-3.3.3-x86_64-3.8/numpy/distutils/include -I/opt/sage-4525b51c9c556e7d1bf7a4d90ee129c6b7f1d9a7/include -I/usr/include/python3.8 -c scipy/stats/_boost/src/beta_ufunc.cxx -o build/temp.cygwin-3.3.3-x86_64-3.8/scipy/stats/_boost/src/beta_ufunc.o -MMD -MF build/temp.cygwin-3.3.3-x86_64-3.8/scipy/stats/_boost/src/beta_ufunc.o.d -std=c++14 -msse -msse2 -msse3" failed with exit status 1
```


### SciPy/NumPy/Python version information

1.7.x

--- a/patches/0001-boost-math-tools-config.hpp-__CYGWIN__-__FreeBSD__-N.patch
+++ b/patches/0001-boost-math-tools-config.hpp-__CYGWIN__-__FreeBSD__-N.patch
@@ -1,0 +1,26 @@
+From 2bc06432b508a2b02026ffa09fe3b08fa28e95f9 Mon Sep 17 00:00:00 2001
+From: Matthias Koeppe <mkoeppe@math.ucdavis.edu>
+Date: Mon, 27 Dec 2021 20:56:55 -0800
+Subject: [PATCH] boost/math/tools/config.hpp [__CYGWIN__, __FreeBSD__]: No
+ longer define BOOST_MATH_NO_LONG_DOUBLE_MATH_FUNCTIONS
+
+---
+ boost/math/tools/config.hpp | 2 +-
+ 1 file changed, 1 insertion(+), 1 deletion(-)
+
+diff --git a/boost/math/tools/config.hpp b/boost/math/tools/config.hpp
+index 480b3e10..057747ef 100644
+--- a/boost/math/tools/config.hpp
++++ b/boost/math/tools/config.hpp
+@@ -28,7 +28,7 @@
+ 
+ #include <boost/math/tools/user.hpp>
+ 
+-#if (defined(__CYGWIN__) || defined(__FreeBSD__) || defined(__NetBSD__) || defined(__EMSCRIPTEN__)\
++#if (defined(__NetBSD__) || defined(__EMSCRIPTEN__)\
+    || (defined(__hppa) && !defined(__OpenBSD__)) || (defined(__NO_LONG_DOUBLE_MATH) && (DBL_MANT_DIG != LDBL_MANT_DIG))) \
+    && !defined(BOOST_MATH_NO_LONG_DOUBLE_MATH_FUNCTIONS)
+ #  define BOOST_MATH_NO_LONG_DOUBLE_MATH_FUNCTIONS
+-- 
+2.35.1
+

